README.md

```diff
@@ -75,260 +75,178 @@ on a template to create a human-readable activity feed. Returns an array of `Act
-
-Form submissions - `Form\Submission::processAction`
-
-GraphQL custom - `GraphQL\CustomAction::onAfterCallMiddleware`
-
-GraphQL general - `GraphQL\GenericAction::afterMutation`
+```yaml
+SilverStripe\Snapshots\Handler\Form\FormSubmissionHandler:
+  messages:
+    publish: 'My publish message'
+```
 
-GridField alter - `GridField\AlterAction::afterCallActionHandler`
+In this case "publish" is the **action identifier** (the function that handles the form).
 
-GridField URL handler - `GridField\UrlHandlerAction::afterCallActionURLHandler`
+### Customising existing snapshot creators
 
-Page `CMSMain` actions - `Page\CMSMainAction::afterCallActionHandler`
+All of the handlers are registered with injector, so the simplest way to customise them is to override their
+definitions in the configuration.
 
-Once you know listener type and the action identifier you need to create an extension which is a subclass of one of the abstract listener handlers.
-Abstract listener depends on your listener type.
+For instance, if you have something custom you with a snapshot when a page is saved:
 
-Form submissions - `Form\SubmissionListenerAbstract`
+```php
+use SilverStripe\Snapshots\Handler\Form\SaveHandler;
+use SilverStripe\Snapshots\Listener\EventContext;
+use SilverStripe\Snapshots\Snapshot;
 
-GraphQL custom - `GraphQL\CustomActionListenerAbstract`
+class MySaveHandler extends SaveHandler
+{
+    protected function createSnapshot(EventContext $context): ?Snapshot
+    {
+        //...
+    }
+}
+```
 
-GraphQL general - `GraphQL\GenericActionListenerAbstract`
+```yaml
+SilverStripe\Core\Injector\Injector:
+  SilverStripe\Snapshots\Handler\Form\SaveHandler:
+    class: MyProject\MySaveHandler
+```
 
-GridField alter - `GridField\AlterActionListenerAbstract`
+### Adding snapshot creators
 
-GridField URL handler - `GridField\UrlHandlerActionListenerAbstract`
+If you have custom actions or form handlers you've added to the CMS, you might want to either ensure their tracked
+by the default snapshot creators, or maybe even build your own snapshot creator for them. In this case, you can
+use the declarative API on `Dispatcher` to subscribe to the events you need.
 
-Page `CMSMain` actions - `Page\CMSMainListenerAbstract`
+Let's say we have a form that submits to a function: `public function myFormHandler($data, $form)`.
 
-**Example implementation**
+```yaml
+SilverStripe\Core\Injector\Injector:
+  SilverStripe\Snapshots\Dispatch\Dispatcher:
+    properties:
+      handlers:
+        -
+          on: [ formSubmitted.myFormHandler ]
+          handler: %$MyProject\Handlers\MyHandler
+```
 
-config
+### Removing snapshot creators
 
-```
-SilverStripe\Snapshots\Snapshot:
-  extensions:
-    - App\Snapshots\Listener\MutationUpdateLayoutBlockGroup
-```
+The configuration API doesn't make it easy to remove items from arrays, so this is best done procedurally.
```

Comment:
Removing a preconfigured event handler doesn't seem to work. It seems that removal of handlers is executed earlier than the population of handlers.

I don't really like this approach as it's runtime configuration. I suggest to use the event name and action pair as the key in the array so we can assign null as the handler. For example:

```
SilverStripe\Core\Injector\Injector:
  SilverStripe\Snapshots\Dispatch\Dispatcher:
    constructor:
      myLoader: [ %$App\Snapshots\Loader ]
    properties:
      handlers:
        'gridFieldAction.handleReorder': null (or %$App\Snapshots\Handler\Reorder for custom handler)
```

Reply:
Yeah, agree this is a weak spot. I had it this way originally, but then it got very verbose when subscribing to multiple actions under the same event.

```
formSubmitted.save: ...
formSubmitted.unpublish: ...
formSubmitted.doSave: ...
formSubmitted.doDelete: ...
```

Which is better suited for a 1:1 relationship between handler and event, but to me, a better model is one handler to many events.

Other option, which wouldn't change the API, is to key the event names:

```
on:
  'formSubmitted.save': true
  'formSubmitted.publish': true
handler:
  MyHandler
```

That way, removal is just

```
on:
  'formSubmitted.save': false
```

Reply:
Have changed this to the keyed events, as above.